module/module.go

```diff
@@ -0,0 +1,501 @@
+		// Expected $class-config.json to exist in ...
+		return nil, errors.Wrapf(
+			err,
+			"Error reading JSON Config %s",
+			classConfigPath,
+		)
+	}
+
+	return &Instance{
+		ClassName:     className,
+		ClassType:     jsonConfig.Type,
+		BaseDirectory: baseDirectory,
+		Directory:     instanceDirectory,
+		InstanceName:  jsonConfig.Name,
+		Dependencies:  readDeps(jsonConfig.Dependencies),
+		JSONFileName:  jsonFileName,
+		JSONFileRaw:   raw,
+	}, nil
+}
+
+func readDeps(jsonDeps map[string][]string) []Dependency {
+	depCount := 0
+
+	for _, depsList := range jsonDeps {
+		depCount += len(depsList)
+	}
+
+	deps := make([]Dependency, depCount)
+	depIndex := 0
+
+	for className, depsList := range jsonDeps {
+		for _, instanceName := range depsList {
+			deps[depIndex] = Dependency{
+				ClassName:    className,
+				InstanceName: instanceName,
+			}
+			depIndex++
+		}
+	}
+
+	return deps
+}
+
+// GenerateBuild will, given a module system configuration directory and a
+// target build directory, run the generators assigned to each type of module
+// and write the generated output to the module build directory
+func (moduleSystem *System) GenerateBuild(baseDirectory string) error {
+	resolvedModules, err := moduleSystem.ResolveModules(baseDirectory)
+
+	if err != nil {
+		return err
+	}
+
+	moduleCount := 0
+	for _, moduleList := range resolvedModules {
+		moduleCount += len(moduleList)
+	}
+
+	moduleIndex := 0
+	for _, className := range moduleSystem.classOrder {
+		classInstances := resolvedModules[className]
+
+		for _, classInstance := range classInstances {
+			moduleIndex++
+			buildPath := filepath.Join(
+				classInstance.BaseDirectory,
+				"build",
+				classInstance.Directory,
+			)
+			fmt.Printf(
+				"Generating %s %s %s in %s %d/%d\n",
+				classInstance.ClassType,
+				classInstance.ClassName,
+				classInstance.InstanceName,
+				buildPath,
+				moduleIndex,
+				moduleCount,
+			)
+
+			classGenerators := moduleSystem.classes[classInstance.ClassName]
+			generator := classGenerators.types[classInstance.ClassType]
+
+			if generator == nil {
+				fmt.Printf(
+					"Skipping generation of %s %s class of type %s as generator is not defined\n",
+					classInstance.InstanceName,
+					classInstance.ClassName,
+					classInstance.ClassType,
+				)
+				continue
+			}
+
+			files, err := generator.Generate(classInstance)
+
+			if err != nil {
+				fmt.Printf(
+					"Error generating %s %s class %s type as generator is not defined\n%s\n",
+					classInstance.InstanceName,
+					classInstance.ClassName,
+					classInstance.ClassType,
+					err.Error(),
+				)
+				return err
+			}
+
+			for filePath, content := range files {
+				filePath = filepath.Clean(filePath)
+
+				if strings.HasPrefix(filePath, "..") {
+					return errors.Errorf(
+						"Module %s generated a file outside the build dir %s",
+						classInstance.Directory,
+						filePath,
+					)
+				}
+
+				resolvedPath := filepath.Join(
```

Comment:
What does `filepath.Join()` do if the second argument is an absolute file path ?